### PR TITLE
Allow host 'celery-0.galaxyproject.eu' to connect to DB 'galaxy'

### PR DESCRIPTION
**Pull request by sj213**
as user 'galaxy' w/ MD5-auth, as requested by bgruening.

### CODE REVIEW

**Review by bgruening — APPROVED** [review]
@sj213 lets try. I think I tested this and it did not work out of the box.

**Comment by sj213** [comment]
Ah, yes, reverse lookups don't work, so this change will probably not help. `celery-0.galaxyproject.eu` can be resolved to `10.5.68.168` but the reverse lookup fails: `Host 168.68.5.10.in-addr.arpa. not found: 3(NXDOMAIN)`
A successful reverse lookup is of course required for using DNS-names in `pg_hba.conf`. This is a problem with our dyndns-setup that is probably not easy to fix. :((

**Comment by bgruening** [comment]
Any other ideas how we can get the IPs and update the playbook automatically maybe?

**Comment by sj213** [comment]
Do you **really** want a script to commit to `infrastructure-playbook/master`, bypassing all checks (assuming it's doable)?

**Comment by bgruening** [comment]
No but maybe creating PRs. I'm worried that the VMS gets out of sync with the DB if we rely on IPs. And something should tell us as soon as possible

**Comment by sj213** [comment]
I've never created a PR programatically, I wouldn't know how to do it (always used the web interface for that). Maybe you should ask a real github-expert. Of maybe even fixing the reverse lookups is easier than that... (Who set up the dyndns for ug.eu in the first place? Helena?)

**Comment by bgruening** [comment]
Yes, I think it was @hexylena. Helena, do you know if this is easy to fix?

**Comment by hexylena** [comment]
> I've never created a PR programatically, I wouldn't know how to do it 

it's easy, there's a github action for it, no worries. You just have a github action make the changes locally, and [then use one of the many PR creation actions.](https://github.com/marketplace/actions/create-pull-request)

> maybe even fixing the reverse lookups is easier than that

Strictly speaking we have no dynamic dns, unless you consider terraform updating routes on a cron job dyndns.
I've never enabled reverse lookups beause I typically didn't rely on them, but it seems pretty easy for AWS.

https://bhargavamin.com/how-to-do/configure-reverse-dns-lookup-amazon-route-53

I.e. we just create the hostedzone

68.5.10.in-addr.arpa.
and add a record
168.68.5.10.in-addr.arpa. PTR celery...

All of this can be done in terraform pretty trivially, and then no automated PR bot needed, and we ensure that if the machine's IP changes, terraform will fix the DNS entries.

**Comment by sj213** [comment]
Creating a DNS reverse zone is IMO the cleanest solution; it's not entirely "clean" as it would, as I understand it, mean advertising non-routable RFC-1918 addresses publicly, which should normally be avoided, but the alternatives available immediately (updating the `pg_hba.conf` file, updating `/etc/hosts`) look even less desireable.

This is assuming Amazon will allow to create reverse-lookup-zones for RFC-1918 addresses, which is by no means understood, as it would pave the way to namespace clashes.

If we wanted a nice and fully best-practices-conforming solution we would, of course, need to put all hosts with 10.x-addresses in a separate DNS zone internal to Freiburg-U, probably on Infoblox (which has an interface for updating zones programatically; the TF is using this for the `informatik.intra.uni-freiburg.de` zone).

**Comment by sj213** [comment]
Come to think of it: We could also use an entirely different approach to limiting DB access to Celery-nodes, e.g. certificate auth:
```
https://www.postgresql.org/docs/13/auth-cert.html
```
This sould mean granting access to DB `galaxy` as user `galaxy` to the entire network `10.5.68.0/24` but *only with a certain trusted SSL cert* - and making sure that the client cert with its privat key is only deployed on the Celery node(s). That said, I've never used cert auth or even plain SSL with PG so I'm not sure how well that would work.

**Comment by bgruening** [comment]
Mh that sounds interesting. Should we try that?

**Comment by sj213** [comment]
We can certainly try the cert approach, but it's a bit involved:
```
https://www.postgresql.org/docs/13/ssl-tcp.html
```
When we want to use client cert auth (with or w/o username verification) we need a small but **automated** CA that can create and sign appropriate client and server certs as well as deploy them via Ansible/Jenkins. Or maybe Jenkins isn't even required as certs need not be renewed more than once a year at most. After all, this will be our very own private CA not for use with HTTPS, we don't have to toe Google's "6-months-absolute-max" line. But of course we must make sure that a valid client cert ist "baked" into every Celery host image.

@hexylena: can you provide a quick pointer to your existing SSH-pubkey mini CA?

Once we have the automated CA going, deploying the generated certs should be easy.

But **before we embark on this route** we should make sure that Celery can actually be configured to use SSL and present a client cert in lieu of a password when connecting to PG, or else the entire effort will have been in vain... XD

**Comment by hexylena** [comment]
https://github.com/usegalaxy-eu/infrastructure-playbook/blob/b7139eb95979af2336c770fe1d4a27db3aae0aef/roles/ssh-host-sign/tasks/main.yml (if I was doing this now, I'd copy the server key locally, sign here, and then send back, but, I was lazy, it only exists on machine, as root, for a matter of seconds.) The invocation is a bit of magic, but, it worked quite well.

tl;dr

- key is stored in ansible (encrypted)
- playbook uses it to sign whatever it needs, and deposit the certificate on the right machine.
- ...
- profit

**Comment by sj213** [comment]
THX @hexylena 

@bgruening: No ansible roles/playbooks just yet (only 3 shell scripts and a config file atm) but I've tried to hack up a Q&D mini-CA in `sn05:/root/CA` and created certificates which can (hopefully) be used for testing the SSL connection and user cert auth. (I think we should try the basic setup to see if it rly works before attempting the full "continuous integration" monty...)

**Comment by bgruening** [comment]
Ok, that sounds great. @mira-miracoli can maybe look at the celery part next week.

Have a nice weekend y'all!

**Comment by mira-miracoli** [comment]
Celery uses Galaxy code and the Galaxy configs, as shown [here](https://github.com/galaxyproject/galaxy/issues/14568#issuecomment-1246420164). There we have a [config.database_connection](https://docs.galaxyproject.org/en/latest/admin/config.html#database-connection)
By quickly consulting stackoverflow, I found this:
```
db_connect_string='mysql://<user>:<pswd>@<db server>:3306/<database>'
ssl_args = {'ssl': {'cert':'/path/to/client-cert', 
                     'key':'/path/to/client-key', 
                      'ca':'/path/to/ca-cert'}}

create_engine(db_connect_string, connect_args=ssl_args)
```
The question is if we have to change a lot of galaxy code in order to be able to use this.

**Comment by hexylena** [comment]
Shouldn't be a *lot* of code, and it would be generally useful to support the full breadth of SSL connections! I guess most of us rely on sockets, or private networks, and having encryption in place is good.

**Comment by bgruening** [comment]
@hexylena do you think you can work with us on the Wiregard setup during the Hackathon in October? Is that something we can learn in 2 days?

**Comment by hexylena** [comment]
Sure, we can do that. Wireguard is extremely simple to setup. I was pointing @natefoo at it as well, maybe he's interested to join too, so, we can all meet together and hack maybe.

If you've got money for tailscale, I really cannot recommend it enough (I know! me, recommending a paid product! It's actually that good.)

**Comment by bgruening** [comment]
I guess we could pay. But we should also consider the EU cloned national Galaxy instances and because of that not depend on payed services? Wireguard seems cool here.

**Comment by hexylena** [comment]
Yeah up to you, just my 2¢; it's just a layer on top of wireguard that handles the (few) annoying points of running the infrastructure (managing keys, authentication, authorization, ACLs, and auditability). Not sure if your new security requirements, require audit logging? but it would solve that rather than requiring writing it yourself.

wireguard is always a cheap place to start though.

**Comment by hexylena** [comment]
> Using Tailscale for an open source or friends & family project? The Community on GitHub plan can get you up to 25 users, 5 devices per user, and 2 admins for free.

ah there is an open source plan, I forgot about this, so there is a way to be cheap @bgruening, I can show all y'all how it works during the cofest and answer more questions then.

**Comment by natefoo** [comment]
We are using the free open source plan now on .org to expose the GxIT kubernetes network to the Galaxy server and it's working flawlessly.

**Comment by mvdbeek** [comment]
> The question is if we have to change a lot of galaxy code in order to be able to use this.

that example would translate to `mysql://<user>:<pswd>@<db server>:3306/<database>?ssl_ca=/path/to/ca-cert&ssl_cert=/path/to/client-cert&ssl_key=/path/to/client-key"` as a connection string, no change to Galaxy required.
